### ISSUE TITLE
Dot access to frontmatter

### ISSUE THREAD

**Issue by joemasilotti** [sitepress/sitepress#16]
I'd love to be able to access frontmatter via dot syntax.

For example, `resource.data.title` instead of `resource.data["title"]`.

**Comment by bradgessler** [comment]
Any opinions on how a `nil` should be handled if it shouldn't be a `nil`?

I've seen frameworks like Middleman handle it via `resource.data.title!`

**Comment by joemasilotti** [comment]
Good question. Bang feels good to me. That leaves the option to check if something exists without any extra syntax.

```ruby
resource.data.title || "Default title"
resource.data.title! # raises
```

**Comment by bradgessler** [comment]
This is currently implemented at https://github.com/sitepress/sitepress/tree/version-4

Here's how it works:

```ruby
resource.data.title || "Default title"
resource.data.title! # Would raise KeyNotFound exception
resource.data.title! "Default title"
resource.data.title! { "Default title" }
```

The `!` mimics the behavior of `Hash#fetch`. Here's the respective Hash equivalents:

```ruby
resource.data.fetch "title", "Default title"
resource.data.fetch "title" { "Default title" }
```

**Comment by joemasilotti** [comment]
Excellent! I can't wait to give this a try.

I don't understand the last example the first block, though. How would something return a string... surrounded by a hash?

**Comment by bradgessler** [comment]
I added the `?` operated to this API to check for the presence of a key that returns a true or a false. For example:

```ruby
resource.data.title? # => true
resource.data.i_do_not_exist? # => false
```

I don't really like this because `dashed-keys-will-not-work`, but it seems to be helpful for Middleman migrations.

I will probably make this more configurable so that this is off by default, but something that can be turned on for people coming in from MM.

**Comment by bradgessler** [comment]
A bigger issue I have to deal with is how to delegate calls from the objects wrapped via `Data.manage` like `includes?`, `keys`, etc. I currently am either using `def_delegators` or implementing a method that provides a similar behavior, but this is effectively like maintaining an "allow list" of methods that will have to be maintained.

Ideally I get something like https://ruby-doc.org/stdlib-2.4.1/libdoc/delegate/rdoc/SimpleDelegator.html in these classes so it mostly Just Works™

**Comment by bradgessler** [comment]
Shipped in 4.0.0
